android/src/test/java/com/sailthru/mobile/rnsdk/JsonConverterTest.kt

```diff
@@ -0,0 +1,346 @@
+package com.sailthru.mobile.rnsdk
+
+import com.facebook.react.bridge.*
+import com.nhaarman.mockitokotlin2.doAnswer
+import com.nhaarman.mockitokotlin2.spy
+import org.json.JSONArray
+import org.json.JSONObject
+import org.junit.Assert.*
+import org.junit.Before
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.junit.runners.BlockJUnit4ClassRunner
+
+@RunWith(BlockJUnit4ClassRunner::class)
+class JsonConverterTest {
+    private val jsonConverter = spy(JsonConverter())
+
+    @Before
+    fun setup() {
+        doAnswer { JavaOnlyMap() }.`when`(jsonConverter).createNativeMap()
+        doAnswer { JavaOnlyArray() }.`when`(jsonConverter).createNativeArray()
+    }
+
+
+    /** JSON -> Map **/
+
+    @Test
+    fun testConvertJsonToMapBoolean() {
+        val jsonObject = JSONObject().apply {
+            put(TEST_KEY, true)
+        }
+
+        val writableMap = jsonConverter.convertJsonToMap(jsonObject)
+
+        assertEquals(true, writableMap.getBoolean(TEST_KEY))
+    }
+
+    @Test
+    fun testConvertJsonToMapInt() {
+        val jsonObject = JSONObject().apply {
+            put(TEST_KEY, 123)
+        }
+
+        val writableMap = jsonConverter.convertJsonToMap(jsonObject)
+
+        assertEquals(123, writableMap.getInt(TEST_KEY))
+    }
+
+    @Test
+    fun testConvertJsonToMapDouble() {
+        val jsonObject = JSONObject().apply {
+            put(TEST_KEY, 1.23)
+        }
+
+        val writableMap = jsonConverter.convertJsonToMap(jsonObject)
+
+        assertEquals(1.23, writableMap.getDouble(TEST_KEY), 0.001)
+    }
+
+    @Test
+    fun testConvertJsonToMapString() {
+        val jsonObject = JSONObject().apply {
+            put(TEST_KEY, "test string")
+        }
+
+        val writableMap = jsonConverter.convertJsonToMap(jsonObject)
+
+        assertEquals("test string", writableMap.getString(TEST_KEY))
+    }
+
+    @Test
+    fun testConvertJsonToMapMap() {
+        val subObject = JSONObject().apply {
+            put("some", "thing")
+        }
+        val jsonObject = JSONObject().apply {
+            put(TEST_KEY, subObject)
+        }
+
+        val writableMap = jsonConverter.convertJsonToMap(jsonObject)
+
+        assertEquals("thing", writableMap.getMap(TEST_KEY)?.getString("some"))
+    }
+
+    @Test
+    fun testConvertJsonToMapArray() {
+        val subArray = JSONArray().apply {
+            put("test string")
+        }
+        val jsonObject = JSONObject().apply {
+            put(TEST_KEY, subArray)
+        }
+
+        val writableMap = jsonConverter.convertJsonToMap(jsonObject)
+
+        assertEquals("test string", writableMap.getArray(TEST_KEY)?.getString(0))
+    }
```

Comment:
is it possible to have a map of mixed types? I know JS allows this, but do these bridge classes? might be worth a test.

Reply:
Yeah, the map can contain whatever mix you like. Good shout, I'll add a more complicated test case 👍